pkg/services/datasources/service/datasource.go

```diff
@@ -252,6 +295,96 @@ func (s *Service) AddDataSource(ctx context.Context, cmd *datasources.AddDataSou
 	})
 }
 
+// This will valid validate the instance settings and mutate the cmd with the processed values
+func (s *Service) mutate(ctx context.Context, pluginContext backend.PluginContext, settings *backend.DataSourceInstanceSettings) (*backend.DataSourceInstanceSettings, error) {
+	operation := backend.AdmissionRequestCreate
+
+	if len(settings.Name) > maxDatasourceNameLen {
+		return nil, datasources.ErrDataSourceNameInvalid.Errorf("max length is %d", maxDatasourceNameLen)
+	}
+
+	if len(settings.URL) > maxDatasourceUrlLen {
+		return nil, datasources.ErrDataSourceURLInvalid.Errorf("max length is %d", maxDatasourceUrlLen)
+	}
+
+	if settings.Type == "" {
+		return settings, nil // NOOP used in tests
+	}
+
+	// Make sure it is a known plugin type
+	p, found := s.pluginStore.Plugin(ctx, settings.Type)
+	if !found {
+		return nil, errutil.BadRequest("datasource.unknownPlugin",
+			errutil.WithPublicMessage(fmt.Sprintf("plugin '%s' not found", settings.Type)))
+	}
+
+	// When the APIVersion is set, the client must also implement ProcessInstanceSettings
+	if p.APIVersion == "" {
+		if settings.APIVersion != "" {
+			return nil, fmt.Errorf("invalid request apiVersion (datasource does not have one configured)")
+		}
+		return settings, nil // NOOP
+	}
+
+	pb, err := backend.DataSourceInstanceSettingsToProtoBytes(settings)
+	if err != nil {
+		return nil, err
+	}
+
+	req := &backend.AdmissionRequest{
+		Operation:     backend.AdmissionRequestCreate,
+		PluginContext: pluginContext,
+		Kind:          settings.GVK(),
+		ObjectBytes:   pb,
+	}
+
+	// Set the old bytes and change the operation
+	if pluginContext.DataSourceInstanceSettings != nil {
+		req.Operation = backend.AdmissionRequestUpdate
+		req.OldObjectBytes, err = backend.DataSourceInstanceSettingsToProtoBytes(settings)
+		if err != nil {
+			return nil, err
+		}
+	}
+
+	rsp, err := s.pluginClient.MutateAdmission(ctx, req)
```

Comment:
why not calling the validation hook first? so not all the plugins needs to call the validation before mutating (and assume that a request to `mutate` may be valid)

Reply:
maybe -- seems like something we can evolve as we have more to work with.  Mutation is the biggest hammer: it takes a payload and does its best to return a valid object.  

Once we have the plumbing in place we can add helpers to improve the process.  At this stage we don't have much.

RIGHT NOW -- before we have simpler options, i think good to require this when "apiversion" is set.   At this stage anyone setting a version is really stepping into risky territory 😜 and asking to participate in the bleeding edge development where support is still evolving

Reply:
I think it's still benefitial to explicitly call the `ValidateAdmission` hook here, seems free and demonstrates how it can be used. Why not adding it? It's not adding any special complexity apart from the one already here.

Reply:
Is this what you have in mind?  If it is helpful, I'm fine adding it...
![image](https://github.com/grafana/grafana/assets/705951/ed83ff1f-0744-4772-90a5-cec4f5389bc8)

It does lose the useful error from mutate, but we can probably add that back to the message (removed when I made is a different payload with only Allowed:bool)

In any case this section will need a lot more attention as we apply this to less trivial things.

Reply:
I think it's useful, yes. At least it's more logical on my mind. My +1 to have the validation hook to return a message too (as long as it does not mutate the input).

Reply:
I added the second block as an example